### PR TITLE
Self argument probing

### PR DESCRIPTION
**Pull request by yuhuishi-convect**
## Description

This adds a static analyzer for the ML engines.
It parses the source code of the ML engine class implementation, 
find how the argument `args` dict is used.

For example, if there is a line `args['some_key']`, then likely there is an argument `some_key` which is required.
It adds the parsed results to the `information_schema.handlers` table.

Next step: when user provides an argument which is not known to the parsed arguments list, we raise a "soft" error to them.

**Fixes** #6846

image

## Type of change

(Please delete options that are not relevant)

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📄 This change requires a documentation update

### What is the solution?

(Describe at a high level how the feature was implemented)

## Checklist:

- [ ] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation, or created issues to update them.
- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] I have shared a short loom video or screenshots demonstrating any new functionality.

### CODE REVIEW

**Review by paxcema — CHANGES_REQUESTED** [review]
Generally looks good to me. There is one bit that I think may be a bit of an edge case and could prove buggy. But otherwise, I believe this will be an extremely useful addition. Nice one @yuhuishi-convect!

File: mindsdb/integrations/libs/base.py [hunk 1]
```diff
@@ -125,7 +133,145 @@ def __init__(self, name: str):
...
+        tree = ast.parse(source_code)
+
+        # find all places where a key in args is accessed
+        # and if it is accessed using args["key"] or args.get("key", "default_value")
+
+        visitor = self.ArgProbeVisitor()
+        visitor.visit(tree)
+
+        # deduplicate the keys
+        unique_arg_keys = set([(r["name"], r["required"]) for r in visitor.arg_keys])
```
[the remark this answers is not in the document]
Comment (yuhuishi-convect) [reply mindsdb/integrations/libs/base.py:238]: Yes nice catch!

File: tests/unit/test_arg_probing.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+def mock_handler_class():
+    class MockHandler(ArgProbeMixin):
+        def __init__(self, **kwargs):
+            ...
+
+        def create(self, args):
+            args["test_required"]
+            args.get("test_optional", "default")
+            args.get("test_optional2")
+            args["test_required2"] = "default"  # this should be ignored
```
[the remark this answers is not in the document]
Comment (yuhuishi-convect) [reply tests/unit/test_arg_probing.py:24]: 👍

**Comment by yuhuishi-convect** [comment]
@paxcema fixed some issues according to the comments

**Review by paxcema — APPROVED** [review]

**Pull request merged**